### PR TITLE
apiDefinition::getRessourceByPath + check var existence and type in some places

### PR DESCRIPTION
**Pull request by DimitriGilbert**
* added apiDefinition::getRessourceByPath to grab resource from the route path
* added check to method::createFromArray to make sure body data is an array (in case of poorly written but still correct RAML)
* use of isset instead of plain check in webFormBody::createFromArray (in case of poorly written but still correct RAML) to avoid fatal error

### CODE REVIEW

File: src/ApiDefinition.php [hunk 1]
```diff
@@ -286,6 +286,37 @@ public function getResourceByUri($uri)
...
+        // get rid of everything after the ?
+        $path = strtok($path, '?');
+
+        $potentialResource = null;
+
+        $resources = $this->getResourcesAsArray($this->resources);
+        foreach ($resources as $resource) {
+            if ($path === $resource->getUri()) {
+                $potentialResource = $resource;
+            }
```
Comment (JoeBengalen) [inline src/ApiDefinition.php]: Why not use an early return here if a match is found? Could improve performance and simplefy the code.

Comment (alecsammon) [inline src/ApiDefinition.php]: Yes - agree - this is very similar to the method above. I think when I wrote the `getResourceByUri` I was keen at the time of having a single return per method, which should always be at the end of the method. However I'm not sure if its a great pattern any more

**Comment by alecsammon** [comment]
Looking good - what are you thoughts on the comment by @JoeBengalen

**Comment by DimitriGilbert** [comment]
That i feel extremely stupid is my thought on @JoeBengalen 's comment as there should anyway be only one (or no) match to a certain path. I'll change that later when i come home from work.

**Comment by alecsammon** [comment]
It was my code originally that was wrong! Stupidity is on my part - not yours!

Feel free to correct my `getResourceByUri` method also

File: src/ApiDefinition.php (hunk 1 above)
Comment (DimitriGilbert) [inline src/ApiDefinition.php]: Yep, late night coding is not always the best... I should have seen that. Anyway...
If you want to keep a single return per method (which i tend to prefer also) why not use `break` to get out of the loop ?
Let me know, I'll update accordingly.

Comment (alecsammon) [inline src/ApiDefinition.php]: It's your choice - whatever you prefer!

Comment (DimitriGilbert) [inline src/ApiDefinition.php]: `break` it is then :)

Comment (JoeBengalen) [inline src/ApiDefinition.php]: If you return as soon as a match is found you can completely remove the `$potentialResource` variable and just throw the exceptions if it reaches the end of the method. You would still have a single `return`.

```php
// get rid of everything after the ?
$uri = strtok($uri, '?');

$resources = $this->getResourcesAsArray($this->resources);
foreach ($resources as $resource) {
    if ($resource->matchesUri($uri)) {
        return $resource;
    }
}

throw new ResourceNotFoundException($uri);
```

Comment (DimitriGilbert) [inline src/ApiDefinition.php]: yep, i did do that in the end

**Comment by alecsammon** [comment]
Sorry for the delay. Merged :)

**Pull request merged**